File: go/2024/01/solution_test.go
package solution

import (
	"log"
	"os"
	"testing"
)

const SampleInput = `3   4
4   3
2   5
1   3
3   9
3   3
`

// Helper function for reading the input file to a string
func readInputFile() string {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		log.Fatalf("Couldn't read input.txt file, %v\n", err)
	}
	return string(data)
}

func TestPart1(t *testing.T) {
	t.Run("sample input", func(t *testing.T) {
		got := Part1(SampleInput)
		want := 11

		assertSolution(t, got, want)
	})
	t.Run("actual input", func(t *testing.T) {
		got := Part1(readInputFile())
		want := 1603498

		assertSolution(t, got, want) // Ravioli, ravioli - give me the formuoli!
	})
}

func TestPart2(t *testing.T) {
	t.Run("sample input", func(t *testing.T) {
		got := Part2(SampleInput)
		want := 31

		assertSolution(t, got, want)
	})
	t.Run("actual input", func(t *testing.T) {
		got := Part2(readInputFile())
		want := 25574739

		assertSolution(t, got, want) // Ravioli, ravioli - give me the formuoli!
	})
}

func assertSolution(t testing.TB, got int, want int) {
	t.Helper()

	if got != want {
		t.Errorf("got %d, wanted %d", got, want)
	}
}

File: python/2024/01/solution_test.py
from solution import part_1, part_2

##################
# Helper Methods #
##################


def get_input(filename: str):
    file = open(filename)
    return file.read()


def run_test(fun, input: str, want: int):
    got = fun(input)
    assert got == want


#########
# Tests #
#########

sample_input = """3    4
4    3
2    5
1    3
3    9
3    3
"""


def test_part_1():
    real_input = get_input("input.txt")

    inputs = {
        "3   7": 4,
        "9   3": 6,
        "2   2": 0,
        sample_input: 11,
        real_input: 0,  # Ravioli, ravioli... GIVE ME THE FORMUOLI!!
    }

    for k, v in inputs.items():
        run_test(part_1, k, v)


def test_part_2():
    real_input = get_input("input.txt")

    inputs = {
        sample_input: 31,
        real_input: 0,  # Ravioli, ravioli... GIVE ME THE FORMUOLI!!
    }

    for k, v in inputs.items():
        run_test(part_2, k, v)
